### DISCUSSION TITLE
Some questions about the construction and the update of the tok2vec layer

### DISCUSSION THREAD

**Discussion by Pandalei97** [explosion/spaCy#8648] [Help: Coding & Implementations]
Hello,

I tried to read the documentations and the source code to understand how Spacy constructs and updates tok2vec layer. I have made a list of questions, some on which I want to have your confirmation, and some on which I still have doubt, and some on which I want to discuss about the best practice. I want to open a thread to gather these questions and answers, in order to help other users and myself to understand better how spacy works. If I have made mistakes in my statements, please feel fine to correct them. Thanks in advance !

## Preambule

Let's make an simple example to better illustrate the questions and answers:

We have trained a static word vector of 128 dimensions.

The config file of the tok2vec:

```
[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v2"
width = ${components.tok2vec.model.encode.width}
attrs = ["ORTH","SHAPE"]
rows = [5000,2500]
include_static_vectors = true

[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
width = 96
depth = 8
window_size = 1
maxout_pieces = 3

```

## Questions

1. The attribute `.tensor` contains the vector calculated by tok2vec/transformer layer. In case of a pipeline that has several components with their independent tok2vec, `.tensor` contains the result of the latest tok2vec ? 

2. In the example, the embedding layer will produce a vector of which the width is 288 (96*3) as the source code shows:

    ```python
   # The source code of Spacy 3.0.6
   concat_size = width * (len(embeddings) + include_static_vectors)
    ```

   And the Maxout component will consider the context and compress the matrix of n*288 into a vector of width 96.

   2.1. The width that we passed for the static vector is 96 while its real size is 128. I see that spacy will do a projection to reduce the dimension. But in another case where tok2vec width is larger than our static vector's width, how does spacy do to project static vectors into a higher dimension space ? Or does it just add 0 in all the additional dimensions ?

   2.2. Are the values in the HashEmbed tables randomly generated ?

   2.3. Is it better to keep tok2vec's width the same as that of our static vectors since the projection may cause information loss ?
3. During the training step, how does spacy update the tok2vec ? Do we just update the weights of the encoder or do we update the HashEmbed tables as well ?
4. If we have already a static vector, by intuition, will it be better not to have too many attributes for MultiHashEmbed layer since it will dilute the importance of static vectors ? 
5. During the training step, I often see the loss of the model and of the tok2vec go easily down to 0.00 like this:
    ```
    ℹ Pipeline: ['tok2vec', 'textcat_multilabel']
    ℹ Initial learn rate: 0.001
    E    #       LOSS TOK2VEC  LOSS TEXTC...  CATS_MACRO_F  SCORE
    ---  ------  ------------  -------------  ------------  ------
      0       0          0.00           0.01         26.20    0.26
      0     200          0.02           3.01         82.66    0.83
      1     400          0.01           0.78         76.92    0.77
      2     600          0.00           0.12         81.48    0.81
      3     800          0.00           0.05         85.05    0.85
      5    1000          0.00           0.03         85.27    0.85
      6    1200          0.00           0.01         85.34    0.85
      9    1400          0.00           0.01         89.43    0.89
     12    1600          0.00           0.00         88.24    0.88
     15    1800          0.00           0.00         89.86    0.90
     .... (other epochs similar)
    ```
    Am I suffering from the vanishing gradient problem ?

**Comment by polm** [comment]
Just to answer one of your questions...

> 5. During the training step, I often see the loss of the model and of the tok2vec go easily down to 0.00 like this:

If your loss goes down and score goes up everything is fine. You might be overfitting, but the network is learning the data you have provided it with. If loss **and** score go down then you could be dealing with a vanishing gradient or other problem.

You don't have a problem, but one thing that might be happening that I've seen before is that the loss scale is a bit off, so it's not really going to zero, it's just got numbers that aren't showing up in the output here. Those could be vanishingly small, or it could be that for whatever reason the scale picked for display here wasn't appropriate. (I've been meaning to look into how that works, as it can be deceptive sometimes.)
